### PR TITLE
Update device models, allow to demo chillers fix/break

### PR DESCRIPTION
**Pull request by dluc**
# Type of change? 

- [x] Bug fix
- [ ] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 

* Update the chiller device models to allow fixing/breaking them during a demo
* Allow methods to suspend the telemetry, e.g. during a reboot
* Refactor the scripting engine, add some more functions to manage the methods execution progress
* Update the device models with the specs provided (add also the Tags, even if they are not supported yet)

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/device-simulation-dotnet/103)

### CODE REVIEW

File: Services/Simulation/JavascriptInterpreter.cs [hunk 1]
```diff
@@ -53,19 +53,28 @@ public class JavascriptInterpreter : IJavascriptInterpreter
             // logging into the service logs
             engine.SetValue("log", new Action<object>(this.JsLog));
 
-            //register callback for state updates
-            engine.SetValue("updateState", new Action<JsValue>(this.UpdateState));
+            // Register callback for state updates
+            engine.SetValue("updateState", new Action<JsValue>(this.JsMergeState));
+            engine.SetValue("reportMethodProgress", new Action<string>(this.JsReportMethodProgress));
```
Comment (timlaverty) [inline Services/Simulation/JavascriptInterpreter.cs]: I'm not sure it's worth changing method progress.  We don't use the method progression reported properties in the twin from the UI right now.  Method progress itself needs redesigned and the sdk support for individual device execution of methods needs evaluated.

File: Services/Simulation/JavascriptInterpreter.cs [hunk 2]
```diff
@@ -53,19 +53,28 @@ public class JavascriptInterpreter : IJavascriptInterpreter
...
 
-            //register sleep function for javascript use
-            engine.SetValue("sleep", new Action<int>(this.Sleep));
+            // Register sleep function for javascript use
+            engine.SetValue("sleep", new Action<int>(this.JsSleep));
+
+            // Register functions to enable/disable internal state simulation
+            engine.SetValue("enableSensorSimulation", new Action(this.JsEnableSensorSimulation));
```
Comment (timlaverty) [inline Services/Simulation/JavascriptInterpreter.cs]: I pictured us turning sensor simulation off at an individual sensor level, not for all sensors.

File: Services/Simulation/JavascriptInterpreter.cs [hunk 3]
```diff
@@ -53,19 +53,28 @@ public class JavascriptInterpreter : IJavascriptInterpreter
...
+            engine.SetValue("disableSensorSimulation", new Action(this.JsDisableSensorSimulation));
+
+            // Register functions to enable/disable telemetry
```
Comment (timlaverty) [inline Services/Simulation/JavascriptInterpreter.cs]: #supernit: "...functions to enable/disable sending telemetry to the hub"

File: Services/Simulation/JavascriptInterpreter.cs [hunk 4]
```diff
@@ -145,14 +154,14 @@ private void JsLog(object data)
             this.log.Debug("Log from JS", () => new { data });
         }
 
-        private void Sleep(int timeInMs)
+        private void JsSleep(int timeInMs)
```
Comment (timlaverty) [inline Services/Simulation/JavascriptInterpreter.cs]: Could be used outside of Js?

File: Services/Simulation/ScriptInterpreter.cs [hunk 5]
```diff
@@ -46,5 +50,40 @@ public class ScriptInterpreter : IScriptInterpreter
...
         }
+
+        public static void AddMethodStatusProperty(Dictionary<string, object> state)
+        {
+            state[DEVICE_METHOD_STATUS] = "";
+        }
+
+        public static void EnableSensorSimulation(Dictionary<string, object> state)
+        {
+            state.Add(SENSOR_SIMULATION_ENABLED, true);
```
Comment (timlaverty) [inline Services/Simulation/ScriptInterpreter.cs]: should you check to see if it exists & set it if yes?  ...adding it if it doesn't

File: Services/data/devicemodels/scripts/EmergencyValveRelease-method.js [hunk 6]
```diff
@@ -2,32 +2,38 @@
...
+
+    disableSensorSimulation();
+
     state.pressure = 150;
-    state.CalculateRandomizedTelemetry = true;
-    // update the state to 150
     updateState(state);
+    log("Pressure decreased to " + state.pressure);
+
+    enableSensorSimulation();
```
Comment (timlaverty) [inline Services/data/devicemodels/scripts/EmergencyValveRelease-method.js]: I don't think you want to enable it again as the simulation calculation will overwrite the pressure value you just set.  I think you want to keep it disabled.

File: Services/data/devicemodels/scripts/IncreasePressure-method.js [hunk 7]
```diff
@@ -22,11 +21,40 @@ var state = {
...
+    sleep(1000);
+
     state.pressure = 250;
-    state.CalculateRandomizedTelemetry = false;
-    // update the state to 250
     updateState(state);
+    log("Pressure increased to " + state.pressure);
+    sleep(1000);
+
+    enableSensorSimulation();
```
Comment (timlaverty) [inline Services/data/devicemodels/scripts/IncreasePressure-method.js]: probably want to leave it disabled so more alarms aren't generated.

File: SimulationAgent/Simulation/DeviceStatusLogic/UpdateDeviceState.cs [hunk 8]
```diff
@@ -108,34 +106,35 @@ private void RunInternal()
...
-                // until the correlating function has been called; e.g. when increasepressure is called, don't write
-                // telemetry until decreasepressure is called for that property.
-                this.log.Debug("Checking for the need to compute new telemetry",
+                // until the correlating function has been called; e.g. when increasepressure is called,
+                // don't write telemetry until decreasepressure is called for that property
+                this.log.Debug("Checking for the need to compute new sensors state",
                     () => new { this.deviceId, deviceState = actor.DeviceState });
-                if ((bool) actor.DeviceState[CALC_TELEMETRY])
+
+                if (ScriptInterpreter.IsSensorSimulationEnabled(actor.DeviceState))
```
Comment (timlaverty) [inline SimulationAgent/Simulation/DeviceStatusLogic/UpdateDeviceState.cs]: it seems odd that the scriptinterpreter would own whether sensor simulation is enabled.  I'd rather keep this in the actor.

Comment (dluc) [reply SimulationAgent/Simulation/DeviceStatusLogic/UpdateDeviceState.cs]: will refactor, clean up will make the PR a little bigger

**Comment by dluc** [comment]
the PR is getting too big and deserves some more cooking time, I will open a new one removing all the cleanup and refactoring

**Pull request closed without merging**